File: 백준/Platinum/11438. LCA 2/LCA 2.py
import sys

sys.setrecursionlimit(1000000)
input = lambda: sys.stdin.readline().rstrip()

def main():
    n = int(input())
    tree = [[] for _ in range(n + 1)]
    for _ in range(n - 1):
        a, b = map(int, input().split())
        tree[a].append(b)
        tree[b].append(a)
    depth = [0] * (n + 1)
    parent = [[0] * 20 for _ in range(n + 1)]
    
    def dfs(x, d):
        depth[x] = d
        for y in tree[x]:
            if not depth[y]:
                parent[y][0] = x
                dfs(y, d + 1)
    
    dfs(1, 1)
    
    for j in range(1, 20):
        for i in range(1, n + 1):
            parent[i][j] = parent[parent[i][j - 1]][j - 1]
    m = int(input())
    
    for _ in range(m):
        a, b = map(int, input().split())
        if depth[a] < depth[b]:
            a, b = b, a
        for i in range(19, -1, -1):
            if depth[a] - depth[b] >= 2 ** i:
                a = parent[a][i]
        if a != b:
            for i in range(19, -1, -1):
                if parent[a][i] != parent[b][i]:
                    a = parent[a][i]
                    b = parent[b][i]
            a = parent[a][0]
        print(a)

if __name__ == '__main__':
    main()

File: 백준/Platinum/11438. LCA 2/LCA 2.java
import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.InputStreamReader;
import java.io.OutputStreamWriter;
import java.util.ArrayList;
import java.util.StringTokenizer;

public class Main {
    static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    static BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
    static StringTokenizer st;
    static int N, M;
    static int[] depth;
    static int[][] parent;
    static ArrayList<Integer>[] tree;

    static void dfs(int node, int d) {
        depth[node] = d;
        for (int i = 0; i < tree[node].size(); i++) {
            int next = tree[node].get(i);
            if (depth[next] == 0) {
                parent[next][0] = node;
                dfs(next, d + 1);
            }
        }
    }

    static void setParent() {
        for (int j = 1; j < 20; j++) {
            for (int i = 1; i <= N; i++) {
                parent[i][j] = parent[parent[i][j - 1]][j - 1];
            }
        }
    }

    static int lca(int a, int b) {
        if (depth[a] > depth[b]) {
            int temp = a;
            a = b;
            b = temp;
        }

        for (int i = 19; i >= 0; i--) {
            if (depth[b] - depth[a] >= (1 << i)) {
                b = parent[b][i];
            }
        }

        if (a == b) return a;

        for (int i = 19; i >= 0; i--) {
            if (parent[a][i] != parent[b][i]) {
                a = parent[a][i];
                b = parent[b][i];
            }
        }

        return parent[a][0];
    }

    public static void main(String[] args) throws Exception {
        N = Integer.parseInt(br.readLine());
        depth = new int[N + 1];
        parent = new int[N + 1][20];
        tree = new ArrayList[N + 1];
        for (int i = 1; i <= N; i++) {
            tree[i] = new ArrayList<>();
        }

        for (int i = 0; i < N - 1; i++) {
            st = new StringTokenizer(br.readLine());
            int a = Integer.parseInt(st.nextToken());
            int b = Integer.parseInt(st.nextToken());
            tree[a].add(b);
            tree[b].add(a);

        }

        dfs(1, 1);
        setParent();

        M = Integer.parseInt(br.readLine());
        for (int i = 0; i < M; i++) {
            st = new StringTokenizer(br.readLine());
            int a = Integer.parseInt(st.nextToken());
            int b = Integer.parseInt(st.nextToken());
            bw.write(lca(a, b) + "\n");
        }
        bw.close();
    }
}